src/util/small_shared_ptr.h

```diff
@@ -0,0 +1,209 @@
+/*******************************************************************\
+
+Module: Small intrusive shared pointers
+
+Author: Reuben Thomas, reuben.thomas@diffblue.com
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_SMALL_SHARED_PTR_H
+#define CPROVER_UTIL_SMALL_SHARED_PTR_H
+
+#include <iosfwd>  // ostream
+#include <utility> // swap
+
+// TODO We should liberally scatter `constexpr`s and `noexcept`s on platforms
+// that support them.
+
+// This class is really similar to boost's intrusive_pointer, but can be a bit
+// simpler seeing as we're only using it one place...
+
+template <typename T>
+class small_shared_ptrt final
+{
+public:
+  small_shared_ptrt()=default;
+
+  explicit small_shared_ptrt(T *t):t_(t)
+  {
+    if(t_)
+    {
+      pointee_increment_use_count(*t_);
+    }
+  }
+
+  small_shared_ptrt(const small_shared_ptrt &rhs):t_(rhs.t_)
+  {
+    if(t_)
+    {
+      pointee_increment_use_count(*t_);
+    }
+  }
+
+  small_shared_ptrt &operator=(const small_shared_ptrt &rhs)
+  {
+    auto copy(rhs);
+    swap(copy);
+    return *this;
+  }
+
+  small_shared_ptrt(small_shared_ptrt &&rhs)
+  {
+    swap(rhs);
+  }
+
+  small_shared_ptrt &operator=(small_shared_ptrt &&rhs)
+  {
+    swap(rhs);
+    return *this;
+  }
+
+  ~small_shared_ptrt()
+  {
+    if(!t_)
+    {
+      return;
+    }
+    pointee_decrement_use_count(*t_);
+    if(pointee_use_count(*t_)==0)
+    {
+      delete t_;
+    }
+  }
+
+  void swap(small_shared_ptrt &rhs)
+  {
+    using std::swap;
+    swap(t_, rhs.t_);
+  }
+
+  T *get() const
+  {
+    return t_;
+  }
+
+  T &operator*() const
+  {
+    return *t_;
+  }
+
+  T *operator->() const
+  {
+    return t_;
+  }
+
+  unsigned use_count() const
+  {
+    return t_?pointee_use_count(*t_):0;
+  }
+
+  explicit operator bool() const
+  {
+    return t_!=nullptr;
+  }
+
+private:
+  T *t_=nullptr;
+};
+
+template <typename T>
+std::ostream &operator<<(std::ostream &os, const small_shared_ptrt<T> &ptr)
+{
+  return os << ptr.get();
+}
+
+template <typename T, typename... Ts>
+small_shared_ptrt<T> make_small_shared_ptr(Ts &&... ts)
```

Comment:
Like `std::make_unique` and `std::make_shared`, this has exception-safety benefits.

In the expression `my_func(small_shared_ptr<X>(new X), small_shared_ptr<Y>(new Y))`, the compiler is allowed to sequence the operations like so:

```
new X
new Y <-- This may throw and leak previously-allocated X
construct small_shared_ptr<X>
construct small_shared_ptr<Y>
my_func
```

Whereas `my_func(make_small_shared_ptr<X>(), make_small_shared_ptr<Y>())` cannot leak because the calls to `make_small_shared_ptr` won't be interleaved.